### DISCUSSION TITLE
Packaging up app sources alongside dependencies and commands.

### DISCUSSION THREAD

**Discussion by c4lliope** [dokku/dokku#5317] [Ideas]
@josegonzalez, I'm digging your rebrand on dokku.com;
color scheme is really oceanic.

I'm considering pre-packaged applications again, and I'm curious -
has anyone made a plugin enabling users to package as a file / load in
an application alongside necessary dependencies?

In my homelab I had been obeying a number of online guides, including:
https://github.com/dionysio/dokku-gitea
https://richardwillis.info/blog/monitor-dokku-server-prometheus-loki-grafana

seems like a normalized zipped folder (`app-name.dokku.gz`?)
could be used as in aid in packaging up all an application's dependencies (config files),
alongside a `dokku ...` command sequence, to basically reproduce a running app on a second machine.
In such cases as a `git` codebase is being used as a basis,
the `.dokku.gz` could include a copy of the repository on a specified hash,
and could include an address usable in `git pull` upgrades.

**Comment by josegonzalez** [comment]
I thought about doing this sort of thing with [omakase](https://github.com/dokku/omakase) - the name is terrible but I don't have a better one atm.

I like ansible for the declarative data-model - I tried to emulate that with the task file (with some additions variable inputs). I also really like terraform's plan/apply model, though I don't know that I have anything for that setup just yet.

I'd like to redo the parsing to be a bit more like [tachyon](https://github.com/vektra/tachyon), after which point it should be easy to iterate over what we have in the ansible-dokku module list, eventually adding support for everything in Dokku (and making ansible-dokku be nothing but a wrapper for it). This would also make it fairly straightforward to send a task file via terraform and more or less add automatic support for terraform as well.

Re: having it run from a repository, that I'm less sure of. We have the `app.json` format that covers some of that, but its a weird balance of not deviating too far from Heroku's `app.json` spec while also making the format useful. I don't really want to add too much there except for common things - I feel like everything else should be driven by IaC-style config application.

**Comment by c4lliope** [comment]
Ah, sure - I also came across https://dokku.github.io/general/the-dokku-experience
and sequel https://dokku.github.io/general/automating-dokku-setup a couple days ago.
Seems like a good summary.
